### DISCUSSION TITLE
aims_merge_sulci_mesh.py Segmentation fault

### DISCUSSION THREAD

**Discussion by sarafv** [brainvisa/brainvisa.github.io#168] [General]
Hello,
I'm getting a Segmentation fault with aims_merge_sulci_mesh.py , in a new install of brainvisa on ubuntu 22, brainvisa-5.1.1.
My cmdline is :

> /home/sara.fernandezvidal/epi_projects/epi_project/thirdPartySoftwareStim/brainvisa/bin/bv python /home/sara.fernandezvidal/epi_projects/epi_project/epiSTIM/share/aims_scripts/aims_merge_sulci_mesh.py /home/sara.fernandezvidal/brainvisaDataBases/epiSTIM/brainvisaDB/patients/pat_ChZi/t1mri/default_acquisition/default_analysis/folds/3.1/Rpat_ChZi.arg sulci.ply

The /home/sara.fernandezvidal/brainvisaDataBases/epiSTIM/brainvisaDB/patients/pat_ChZi/t1mri/default_acquisition/default_analysis/folds/3.1/Rpat_ChZi.arg is fine

Thanks for your help!
Sara

**Comment by denisri** [comment]
Hi,
I don't know a script `aims_merge_sulci_mesh.py`, it seems to be a script of yours, so I can't really tell...
If you can share the script we could perhaps try it...
Denis

**Reply by sarafv** [reply]
hi,
sorry, this is the code that woks well until the last install :

```
> > # This script extracts sulci mesh from the output of the BrainVISA Morphologist pipeline.
> > # The scripts take three argiments:
> > #   1: left sulci graph
> > #   2: right sulci graph
> > #   3: path of the output mesh
> > 
> > import sys,os
> > from soma import aims
> > 
> > 
> > def mergeMeshedFromSulciGraph(graph):
> >   """
> >   Merges meshes from a single graph
> >   """
> >   mergedMesh = None
> >   
> >   # iterate on vertices
> >   for vertex in graph.vertices():
> >       try:
> >           # access mesh in attribute 'aims_Tmtktri'
> >           
> >           mesh = vertex[ 'aims_Tmtktri' ]
> >           if mergedMesh == None:
> >               mergedMesh = mesh
> >           else:
> >               print("before merge mesh")
> >               print(vertex[ 'skeleton_label' ])
> >               aims.SurfaceManip.meshMerge(mergedMesh,mesh)
> >       except:
> >           pass
> >   return mergedMesh
> > 
> > 
> > def createMergedMesh(graphFile,meshOutFile):
> >   """
> >   Creates merged sulci meshe
> >   """
> >   graph= aims.read(graphFile)
> >   mesh = mergeMeshedFromSulciGraph(graph)
> >   aims.write(mesh,meshOutFile,options={'ascii' : True})
> > 
> > 
> > if __name__ == '__main__':
> > 
> >   graphFile = sys.argv[1]
> >   meshOutFile = sys.argv[2]
> > 
> >   # create mesh
> >   createMergedMesh(graphFile,meshOutFile)
> >

```

The segmentation fault occurs at  :

aims.SurfaceManip.meshMerge(mergedMesh,mesh)



///. 
before merge mesh
550
Segmentation fault (core dumped)

**Comment by denisri** [comment]
Hi,
I can reproduce the problem, and I have spent some time debugging it. It happens that for an unknown reason a call to a python function (in C) sometimes fails and then works again when called a second time, which seems completely crazy.
Anyway I could work around the problem, but in the development source codes.
To fix your script for now, you can replace the lines:
```
        if mergedMesh == None:
            mergedMesh = mesh
        else:
            print("before merge mesh")
            print(vertex[ 'skeleton_label' ])
            aims.SurfaceManip.meshMerge(mergedMesh,mesh)
```
with:
```
        if mergedMesh == None:
            mergedMesh =  type(mesh._get())(mesh._get())   ## 1
        else:
            print("before merge mesh")
            print(vertex[ 'skeleton_label' ])
            aims.SurfaceManip.meshMerge(mergedMesh, mesh._get())  ## 2
```
change ## 1 is just to copy the first mesh into a new object, otherwise you will be modifying the mesh of the first graph vertex. It's maybe not important in this specific context, but could be bad if you re-use the sulci graph later in the program.
change ## 2 just manually calls the conversion from a wrapper smart-pointer in C++ to the underlying object. This should be automatic but is what is not working here.
Denis

**Reply by sarafv** [reply]
Thank you very much. I'll test it tomorrow
Sara

**Comment by sarafv** [comment]
It is working well,
Thanks!
